# Homework 1 - NL&IP

## Preparing the Data

Importing the modules

In [2]:
import pandas as pd
import numpy as np

from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.feature_selection import SelectFromModel
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_validate

Retreiving the data from file and showing first elements

In [4]:
dataframe=pd.read_csv("dataset_simple.csv")
dataframe.head(10)

feature_0  feature_1  feature_2  feature_3  feature_4  feature_5  \
0  -0.272674   0.801643  -0.650531   0.952486  -0.153037   0.427145   
1   0.201004  -0.263537  -0.793137  -4.483672   0.745123   1.156577   
2  -1.834495  -0.305209   0.053789   4.448289   1.294481   1.241900   
3  -0.484526   0.473056  -0.345463  -7.375136  -0.817676  -0.778446   
4  -0.032351   0.247055  -0.351119  -2.669246  -1.068889   0.322954   
5   0.611700  -0.244822   0.583607   1.503404  -0.104725   0.943069   
6  -0.683728   0.624575  -2.098098  14.415604   0.089336   0.591611   
7   0.522493  -1.551632   1.037685   4.540567  -2.180189  -1.091859   
8   0.722775   1.323710  -0.487995  -6.635840   0.371237  -0.643517   
9  -0.086277  -0.138011   0.488998  -6.420821   0.529817  -0.559980   

   feature_6  feature_7  feature_8  feature_9  ...    feature_991  \
0  -1.085534   1.318513   0.002588  -1.849832  ...      -0.572848   
1   0.351558   1.735511  -1.490988  -0.327063  ...       1.045496   
2  -1.358354  -0.211469  -0.186544   0.964812  ...       0.917945   
3   0.909734  -0.677147  -1.913900  -0.087426  ...       0.598949   
4   1.101136   1.817973   0.944537   0.406005  ...      -0.851188   
5   1.826918  -0.694449  -0.113367   0.909642  ...       0.441221   
6   0.292786   0.853558   2.063484  -1.549264  ...       1.298536   
7  -0.155177   1.039134  -2.164338  -1.310618  ...       0.263004   
8   0.402858  -0.378908  -1.435613  -1.843280  ...       0.977186   
9  -0.089184  -0.597874   0.780693   0.626407  ...      -2.018879   

   feature_992  feature_993  feature_994  feature_995  feature_996  \
0    -0.448690     0.073369    -0.845503    -1.074903     0.222103   
1     1.494638    -0.776573    -1.219524    -0.756398    -0.963046   
2     0.920003     0.690022    -0.834381    -0.261806    -0.061880   
3     0.078195     0.024128     1.308711     0.943230     0.903181   
4     1.272217     0.172442    -0.509007     0.357073     0.523263   
5    -1.386731     2.167706     0.143728    -0.031690    -0.710304   
6     0.848869    -2.538228     1.595099    -0.365556     2.755379   
7     1.851860     0.440721    -0.005803     1.211405     0.489479   
8     1.137189    -0.331244     1.068881     1.736115     0.171098   
9    -1.184220     1.086613     0.149949    -0.265005    -0.370810   

   feature_997  feature_998  feature_999  label  
0    -1.024658    -0.711647     0.036411      1  
1     0.488610     0.313609    -0.490794      0  
2    -1.177730    -1.345305    -0.085971      0  
3     0.501013     2.639802     1.540640      1  
4     0.965078     0.297335     0.088714      1  
5    -0.406994     1.284103     0.398134      1  
6    -1.017768    -0.028076     0.098568      0  
7     0.868569     0.726896     1.076512      1  
8     1.582741    -2.111628    -1.630762      1  
9     0.314975     0.601241    -0.272400      1  

[10 rows x 1001 columns]

Desribing the data

In [5]:
dataframe.describe()

feature_0    feature_1    feature_2    feature_3    feature_4  \
count  9999.000000  9999.000000  9999.000000  9999.000000  9999.000000   
mean     -0.004460    -0.013556    -0.003732    -0.526323     0.005371   
std       0.991939     1.004086     1.004643     5.892651     0.995250   
min      -3.769974    -3.617617    -3.753391   -21.830355    -3.952088   
25%      -0.667193    -0.702984    -0.666734    -4.535838    -0.664072   
50%       0.006990    -0.019359     0.002527    -0.596745     0.008596   
75%       0.655618     0.667009     0.660033     3.367154     0.672424   
max       3.629974     4.437563     3.724971    20.021875     3.444207   

         feature_5    feature_6    feature_7    feature_8    feature_9  \
count  9999.000000  9999.000000  9999.000000  9999.000000  9999.000000   
mean     -0.008539     0.017156    -0.001695     0.016588    -0.011591   
std       1.001142     0.995826     1.000576     1.016175     0.994338   
min      -4.302776    -3.693491    -3.769295    -3.582291    -3.964497   
25%      -0.676919    -0.657419    -0.678497    -0.670545    -0.696125   
50%      -0.014751     0.023373    -0.005377     0.007411    -0.023052   
75%       0.662462     0.697545     0.662622     0.720524     0.668373   
max       3.502859     3.925565     4.069058     5.209648     3.871048   

          ...       feature_991  feature_992  feature_993  feature_994  \
count     ...       9999.000000  9999.000000  9999.000000  9999.000000   
mean      ...          0.004245     0.004588     0.000104    -0.004114   
std       ...          1.010292     0.996613     0.984661     1.005568   
min       ...         -3.543944    -3.702521    -3.645616    -3.757164   
25%       ...         -0.686178    -0.674913    -0.663270    -0.676386   
50%       ...         -0.002653     0.000651    -0.005459    -0.002117   
75%       ...          0.682821     0.680257     0.668761     0.675716   
max       ...          3.616481     4.456758     3.727838     3.959971   

       feature_995  feature_996  feature_997  feature_998  feature_999  \
count  9999.000000  9999.000000  9999.000000  9999.000000  9999.000000   
mean     -0.015524     0.003792    -0.012826    -0.000154     0.008867   
std       1.000405     0.992878     0.986676     1.003824     0.992667   
min      -3.998032    -3.844831    -4.314576    -3.912707    -3.852604   
25%      -0.690162    -0.660690    -0.671733    -0.689658    -0.653306   
50%      -0.017218     0.005911    -0.009118     0.024641    -0.002785   
75%       0.660026     0.680975     0.647070     0.674688     0.677321   
max       3.520052     4.305980     3.671758     3.555786     3.574397   

             label  
count  9999.000000  
mean      0.493149  
std       0.499978  
min       0.000000  
25%       0.000000  
50%       0.000000  
75%       1.000000  
max       1.000000  

[8 rows x 1001 columns]

Ensuring data set is not assymetric

In [6]:
dataframe.shape
dataframe['label'].value_counts()

0    5068
1    4931
Name: label, dtype: int64

Getting x and y and getting rid of label entries in x

In [7]:
x_matrix = dataframe.copy()
x_matrix.drop(['label'], axis=1, inplace=True)
y_vector = dataframe['label']

Normalizing data set

In [8]:
sc = StandardScaler()
sc.fit(x_matrix)
x_matrix = sc.transform(x_matrix)

## Feature selection

In [9]:
classif = ExtraTreesClassifier(n_estimators=100)
classif = classif.fit(x_matrix, y_vector)
classif.feature_importances_  
selected = SelectFromModel(classif, prefit=True)
x_matrix_new = selected.transform(x_matrix)
x_matrix_new.shape

(9999, 147)

Splitting data into desired samples

In [10]:
X_train, X_test, y_train, y_test = train_test_split(x_matrix_new, y_vector, test_size=0.33, random_state=42,shuffle=True)

## Getting classifiers for three models

In [11]:
classif_LR = LogisticRegression()
classif_KN = KNeighborsClassifier()
classif_RF = RandomForestClassifier()

## Retreiving the models on training sample 

### Loglikelihood

In [12]:
classif_LR.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

### KNearestNeighbour

In [16]:
classif_KN.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=5, p=2,
           weights='uniform')

### Random Forest

In [17]:
classif_RF.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

## Cross Validation

### Loglikelihood

In [19]:
scores_LR = cross_validate(estimator=classif_LR, X=X_train, y=y_train, cv=10, scoring='roc_auc')

print('AUC Loglikelihood: ', scores_LR['test_score'])
print('Mean AUC: ', np.mean(scores_LR['test_score']))
print("")

AUC Loglikelihood:  [0.90794304 0.91471521 0.9075777  0.90039564 0.9239022  0.89741945
 0.90926183 0.91531223 0.90404904 0.89912414]
Mean AUC:  0.9079700487085531



### KNearestNeighbour

In [20]:
scores_KN = cross_validate(estimator=classif_KN, X=X_train, y=y_train, cv=10, scoring='roc_auc')

print('AUC KNeighbours: ', scores_KN['test_score'])
print('Mean AUC: ', np.mean(scores_KN['test_score']))
print("")

AUC KNeighbours:  [0.98650021 0.99349515 0.98882146 0.98128297 0.98338591 0.98369333
 0.98905314 0.97939389 0.98410322 0.983238  ]
Mean AUC:  0.9852967293025887



### Random Forest

In [21]:
scores_RF = cross_validate(estimator=classif_RF, X=X_train, y=y_train, cv=10, scoring='roc_auc')

print('AUC Random Forest: ', scores_RF['test_score'])
print('Mean AUC: ', np.mean(scores_RF['test_score']))
print("")

AUC Random Forest:  [0.78567419 0.8262181  0.79267804 0.82247113 0.80495259 0.81884
 0.8143267  0.81613559 0.81821179 0.81372777]
Mean AUC:  0.8113235894947843



## Conclusion

Cross valuation shows that KNearestNeigbour is the most accurate model when dealing with this data. Its area under the curve score is 98.5%, significantly better than the two other models. 